draft-ietf-jose-hpke-encrypt.md

```diff
@@ -131,70 +131,172 @@ This specification uses the following abbreviations and terms:
 
-# Overview
+# JOSE-HPKE
 
-This specification describes two modes of use for HPKE in JWE:
+This document specifies two modes for JOSE-HPKE:
 
-  *  HPKE JWE Integrated Encryption, where HPKE is used to encrypt the plaintext.
-  *  HPKE JWE Key Encryption, where HPKE is used to encrypt a content encryption key (CEK) and the CEK is subsequently used to encrypt the plaintext.
+  *  HPKE JWE Integrated Encryption, where HPKE is used to directly encrypt the plaintext.
+  *  Key Encryption, where HPKE is used to encrypt a content encryption key (CEK) for encrypting the plaintext.
 
-When "alg" and "enc" are both present in a protected header and when "iv" and "tag" are empty, the mode is HPKE JWE Integrated Encryption.
+When "alg" is a JOSE-HPKE algorithm:
 
-When "enc" is present in a protected header and "alg" is absent, the mode is HPKE JWE Key Encryption when a valid HPKE "alg" value is present in the unprotected headers.
+  * If "enc" is "dir", HPKE JWE Integrated Encryption is used.
+  * If "enc" is an AEAD algorithm, The recipient Key Managment mode is Key Encryption.
+
+The HPKE KEM, KDF and AEAD used depend on the JOSE-HPKE algorithm used.
 
 HPKE supports several modes, which are described in Table 1 of {{RFC9180}}.
 
-In JWE, the use of specific HPKE modes such as "mode_base" or "mode_auth_psk" is determined by the presence of the header parameters "psk_id" and "auth_kid".
+The HPKE mode used is determined by header parameters "psk_id" and "auth_kid" in the JOSE Header:
+
+  * If neither is present, base mode is used.
+  * If only "psk_id" is present, PSK mode is used.
+  * If only "auth_kid" is present,  Auth mode is used.
+  * If both are present, AuthPSK mode is used.
+
+If authentication is not supported by the HPKE KEM used, the "auth_kid" header parameter MUST be absent.
+
+
+## HPKE JWE Integrated Encryption
+
+When encrypting, the inputs to HPKE Seal operation are set as follows:
+
+- kem_id: Depends on the JOSE-HPKE algorithm used.
+- pkR: The recipient public key, converted into HPKE public key.
+- kdf_id: Depends on the JOSE-HPKE algorithm used.
+- aead_id: Depends on the JOSE-HPKE algorithm used.
+- info: By default, an empty string. Application MAY specify some other value.
+- aad: Additional Authenticated Data encryption parameter defined in {{RFC7516}} section 5.1 step 14.
+- pt: The message plaintext, compressed using the algorithm in "zip" header parameter if present.
+
+Additionally, if the "psk_id" header parameter is present:
+
+- psk: The pre-shared key.
+- psk_id: The value of the "psk_id" header parameter.
+
+Additionally, if the "auth_kid" header parameter is present:
+
+- skS: The sender private key, converted into HPKE private key.
+
+The resulting JWE is filled as follows:
+
+- JWE Protected Header:
+  * SHALL contain "enc" with value "dir"
+  * SHALL contain "alg" that is the used JOSE-HPKE algorithm.
+  * MUST contain the "apu", "apv" and "zip" header parameters, if present.
```

Comment:
Replace all occurrences of "SHALL" with "MUST" (lines 187 to 192).

Reply:
Thinking about it, switching between keywords with no defined or obvious pattern (but in this case there is a pattern) looks odd.

Reply:
Yes, while "SHALL" and "MUST" both indicate mandatory requirements, I prefer sticking to one term in the document which would improve consistency and clarity.